src/ops/mul_add.rs

```diff
@@ -0,0 +1,143 @@
+/// The fused multiply-add operation.
+/// Computes (self * a) + b with only one rounding error.
+/// This produces a more accurate result with better performance
+/// than a separate multiplication operation followed by an add.
+///
+/// Note that `A` and `B` are `Self` by default, but this is not mandatory.
+///
+/// # Example
+///
+/// ```
+/// use std::f32;
+///
+/// let m = 10.0_f32;
+/// let x = 4.0_f32;
+/// let b = 60.0_f32;
+///
+/// // 100.0
+/// let abs_difference = (m.mul_add(x, b) - (m*x + b)).abs();
+///
+/// assert!(abs_difference <= f32::EPSILON);
+/// ```
+pub trait MulAdd<A = Self, B = Self> {
+    /// The resulting type after applying the fused multiply-add.
+    type Output;
+
+    /// Performs the fused multiply-add operation.
+    fn mul_add(self, a: A, b: B) -> Self::Output;
+}
+
+/// The fused multiply-add assignment operation.
+pub trait MulAddAssign<A = Self, B = Self> {
+    /// Performs the fused multiply-add operation.
+    fn mul_add_assign(&mut self, a: A, b: B);
+}
+
+impl MulAdd<f32, f32> for f32 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_add(self, a: Self, b: Self) -> Self::Output {
+        #![allow(unconditional_recursion)]
+        f32::mul_add(self, a, b)
+    }
+}
+
+impl MulAdd<f64, f64> for f64 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_add(self, a: Self, b: Self) -> Self::Output {
+        #![allow(unconditional_recursion)]
+        f64::mul_add(self, a, b)
```

Comment:
I believe this `allow` is masking a real problem.  In a `#![no_std]` build, `f64` doesn't have most of its inherent methods, which means this call will resolve back to `MulAdd::mul_add` again -- thus the unconditional recursion.

Reply:
Oh `#![no_std]` is what made the builds fail on CI, yet succeed on my local machine.
Adding the `#![allow(unconditional_recursion)]` felt fishy.

How would you propose solving this? Something like this?

```rust
if cfg!(feature == "std") {
    f64::mul_add(self, a, b)
} else {
    (self * a) + b
}
```

Reply:
I think it would be better to *not* implement the trait at all for floats in `#![no_std]` mode, so we don't mislead about the rounding errors.  Either that, or do a manual implementation with appropriate rounding fixups, but I imagine that's non-trivial.